### ISSUE TITLE
Swift 3 / Xcode Beta 6

### ISSUE THREAD

**Issue by dbburgess** [badoo/Chatto#214]
Chatto is a fantastic resource, thank you for all of your work on it and for open sourcing it. 👍 

I see there is an existing Swift 3 branch, which appears to be up to date with beta 5. With beta 6 out, and presumably, an impending official release soon, we're looking to migrate our project to Swift 3 over the next week or two in preparation for the final release to support some new APIs and features along with (or soon after) the iOS 10 launch.

What are your plans for the swift-3 branch? Is there a plan / timetable for the migration? Also, it doesn't appear to be up-to-date with the latest 2.0 release, either. We're happy to lend a hand and help out, but didn't want to jump in if you had specific plans or were already actively working on supporting beta 6 / version 2.0. Just let me know how best we can help!

**Comment by diegosanchezr** [comment]
swift-3 branch is experimental, just to see how the migration process, play with Swift 3 and find/report bugs in the compiler.

The plan is to throw it away (meaning start with the migration again from scratch with Xcode 8 GM or later - hopefully it will be more polished -). Now it's just a PoC with little review of what the migrator did. We need to carefully review all the changes. More important ones:

- [ ] Audit label of first parameter. Sometimes the migrator removes it when it's not what we want
- [ ] Audit open/public access modifiers.
- [ ] Decide what we do about SE-0111 and the lost of labels in closures at call site (I'm in favor doing [this](https://twitter.com/spaxeon/status/765330050233298944))

The schedule for this is tightly coupled to the adoption of Swift 3 at Badoo. I can't give you exact dates, but we want to do it ASAP.

Remember that Chatto supports Swift 2.3 already, so you can use it with Xcode 8 and iOS 10 SDK today.

**Comment by diegosanchezr** [comment]
Anyways I've updated the branch to compile with swift-DEVELOPMENT-SNAPSHOT-2016-08-31-a.xctoolchain. It won't compile with the version bundled in Xcode Beta 6 due to  [SR-2355](https://bugs.swift.org/browse/SR-2355)

**Comment by weyert** [comment]
Thank you, Diego!

**Comment by dbburgess** [comment]
Thanks for the update, @diegosanchezr. Given the continued instability of 3.0, we're just using 2.3 for now. It's working fine. Hopefully 3.0 smooths out soon!

**Comment by ogezue** [comment]
Currently Chatto is the only app that needs to be updated to Swift 3.0 in my library stack. What do you think when a Swift 3.0 will be available?

**Comment by irace** [comment]
@diegosanchezr: What branch or tag is implemented using Swift 2.3? I'm having trouble finding it...

**Comment by dbburgess** [comment]
@irace: There have been commits to add compatibility for Swift 2.3 / Xcode 8, but I don't believe there has been an official release since then with support yet. If you just use what's currently on the `master` branch, it should work fine. It worked for me, anyway.

**Comment by makoni** [comment]
+1 for Swift 3 update

**Comment by diegosanchezr** [comment]
Update: 
Chatto 2.1 was released 2 days ago, supports Xcode 8 + Swift 2.3

swift-3 branch is up to date with master, PR is created an tests are passing. I've reduced the use of open access level maybe too much, feel free to give it a try and complain in the PR if you're not able to override something. The plan is to merge it this week.

swift-2 branch was created. Once swift-3 branch is merged we'll support both Swift 2 and 3 during a short time. This means that once swift-3 branch is in master we'll release Chatto 3.0 and all PRs should be duplicated against master and swift-2 branch until we finally drop support for Swift 2
